### PR TITLE
Fix the misleading duplicate CloudEvent failure under concurrent DCB appends

### PR DESCRIPTION
**Pull request by johanhaleby**
## What

I fixed concurrent DCB appends failing with a misleading duplicate CloudEvent error on a MongoDB event store that has both the DCB and STREAM capabilities enabled (the default agnostic configuration).

A DCB append is placed in a partition stream by the partitioned stream-id generator. Two appends to disjoint DCB boundaries can hash to the same partition and race on the next stream version, and the loser hits the unique index over `streamid` and `streamversion`. That collision is not a genuine duplicate CloudEvent, because the events carry unique `id` and `source` values, but it was mistranslated to a non-retryable `DuplicateCloudEventException`. The append then failed with a misleading error instead of being retried.

## How

I added `MongoExceptionTranslator.isDuplicateKeyErrorOnStreamVersionIndex`, which tells a duplicate-key error on the `streamid` plus `streamversion` index apart from one on the `id` plus `source` index by inspecting the duplicate-key message. The DCB insert paths in the native store, the Spring blocking store, and the Spring reactor store now rethrow the raw duplicate-key error for a stream-version collision, so the existing retry loop reruns the read, decide, and append cycle. A genuine `id` plus `source` duplicate still becomes a `DuplicateCloudEventException`. I did not broaden the retry predicate to treat every `DuplicateCloudEventException` as retryable, so a real duplicate stays a hard failure. I also corrected the code comments that described the opposite behavior.

## Honest note on reproducibility

On the pinned MongoDB version 4.2.8, the collision inside the append transaction surfaces as a transient write conflict that `ClientSession.withTransaction` already retries on its own, so the misleading `DuplicateCloudEventException` path is not reached there. I verified this across many thousands of concurrent append races at default and snapshot read concern, and none failed. This is therefore not a confirmed live bug on 4.2.8. The fix corrects the exception translation and the misleading comments, and it defends the non-transient duplicate-key path that other MongoDB versions and configurations can return for the loser rather than a transient write conflict.

Recommended follow-up: confirm reproduction on MongoDB 5.0 or later under a real concurrent load.

## Tests

Deterministic, MongoDB-version-independent gate:

1. `MongoExceptionTranslatorTest` asserts `isDuplicateKeyErrorOnStreamVersionIndex` returns true for a synthesized duplicate-key error on the `streamid` plus `streamversion` index and false for one on the `id` plus `source` index, and that the latter still translates to `DuplicateCloudEventException`.
2. `MongoEventStoreDcbConcurrencyTest.non_transient_stream_version_duplicate_is_retried_to_success` wraps the event collection so the first `insertMany` in the DCB append path throws a non-transient duplicate-key error on the stream-version index, then asserts the append is retried and commits. Against the unmodified code that injected error becomes a non-retryable `DuplicateCloudEventException` and the append fails. After the fix it is retried and the event commits.

I also kept `single_partition_disjoint_boundaries_are_retried_to_success` as a regression guard. It forces every append into one partition and asserts all concurrent appends commit. It passes before and after on 4.2.8.

Fail-before, against unmodified source with the fix stashed:

```
non_transient_stream_version_duplicate_is_retried_to_success <<< ERROR!
DuplicateCloudEventException[id='null', source=null, details='Bulk write operation error ...
index: streamid_1_streamversion_1 dup key: { streamid: "dcb:partition:", streamversion: 1 }']
	at MongoExceptionTranslator.translateToDuplicateCloudEventException(MongoExceptionTranslator.java:80)
```

Pass-after, same test with the fix applied:

```
MongoExceptionTranslatorTest      Tests run: 3, Failures: 0, Errors: 0
MongoEventStoreDcbConcurrencyTest Tests run: 12, Failures: 0, Errors: 0
```

Full verification across the four affected modules (native, common, Spring blocking, Spring reactor): 475 tests, 0 failures, 0 errors.

### CODE REVIEW

File: eventstore/mongodb/common/src/main/java/org/occurrent/eventstore/mongodb/internal/MongoExceptionTranslator.java [hunk 1]
```diff
@@ -61,6 +62,31 @@ public static RuntimeException translateException(WriteContext ctx, MongoExcepti
...
+                    .anyMatch(bulkWriteError -> referencesStreamVersion(bulkWriteError.getMessage()));
+        }
+        return e.getCode() == 11000 && referencesStreamVersion(e.getMessage());
+    }
+
+    private static boolean referencesStreamVersion(String message) {
+        // The duplicate-key message names the offending index and its key fields. The stream-version index message names
+        // the streamversion field, the id+source index message does not, so this distinguishes the two.
+        return message != null && message.contains(OccurrentCloudEventExtension.STREAM_VERSION);
+    }
```
[the remark this answers is not in the document]
Comment (johanhaleby) [reply eventstore/mongodb/common/src/main/java/org/occurrent/eventstore/mongodb/internal/MongoExceptionTranslator.java:98]: Good catch. I switched `referencesStreamVersion` from a bare `contains("streamversion")` to matching the offending index by name, `index: streamid_1_streamversion_1`, built from `OccurrentCloudEventExtension.STREAM_ID` and `STREAM_VERSION` so it tracks the actual index name. I also added a regression test whose `source` value contains `streamversion`, locking in that an id and source duplicate is no longer misclassified as a retryable stream-version collision.

File: eventstore/mongodb/native/src/test/java/org/occurrent/eventstore/mongodb/nativedriver/MongoEventStoreDcbConcurrencyTest.java [hunk 2]
```diff
@@ -634,6 +719,33 @@ private MongoEventStore buildEventStoreWithStreamIdGenerator(DcbStreamIdGenerato
...
+        AtomicInteger insertManyCalls = new AtomicInteger(0);
+        return (MongoCollection<Document>) Proxy.newProxyInstance(
+                MongoCollection.class.getClassLoader(),
+                new Class<?>[]{MongoCollection.class},
+                (proxy, method, args) -> {
+                    if ("insertMany".equals(method.getName()) && insertManyCalls.getAndIncrement() == 0) {
+                        throw streamVersionDuplicateKeyException();
+                    }
+                    return method.invoke(realCollection, args);
+                });
```
[the remark this answers is not in the document]
Comment (johanhaleby) [reply eventstore/mongodb/native/src/test/java/org/occurrent/eventstore/mongodb/nativedriver/MongoEventStoreDcbConcurrencyTest.java:744]: Fixed. The proxy now catches `InvocationTargetException` and rethrows its cause, so the injected exception reaches the code under test with its real type instead of the reflection wrapper. That keeps the deterministic injected-error gate faithful to the real duplicate and transient classification path.

File: changelog.md [hunk 3]
```diff
@@ -63,6 +63,11 @@ DCB is a capability layered on the existing CloudEvent storage, not a new store
   auto-configure the `DcbApplicationService` even when no global `TagGenerator` bean exists, so a decider-only
   application needs none, while `@DcbTag` and raw-execute users still get a global tagger when one is present.
   * See [ADR 52](doc/architecture/decisions/0052-couple-decider-with-dcb-boundary-and-tags.md).
+* Fixed concurrent DCB appends failing with a misleading duplicate CloudEvent error on a store with both the DCB and
+  STREAM capabilities enabled. Two appends to disjoint DCB boundaries can hash to the same partition stream and race on
+  the next stream version, and the loser hit the unique `streamid` plus `streamversion` index. That collision is not a
```
[the remark this answers is not in the document]
Comment (johanhaleby) [reply changelog.md]: Done. I reworded the entry to present tense, so it reads `the loser hits the unique streamid plus streamversion index` rather than `hit`.
